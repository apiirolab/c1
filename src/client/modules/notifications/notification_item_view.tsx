import * as React from 'react';

import { QueryTypes } from 'data/client';

import { Link } from '@reach/router';
import { inject } from 'mobx-react';
import { Icon, List } from 'semantic-ui-react';
import styled from 'styled-components';

import { RemoveNotificationMutation } from './notifications_queries';

// dayjs.extend(relativeTime);

const Time = styled.span`
  color: #aaa;
`;

const ClickableIcon = styled(Icon)`
  cursor: pointer;
`;

type Props = {
  notification: QueryTypes.Notifications;
  context?: App.Context;
};

export const NotificationView = inject('context')(({ notification, context }: Props) => (
  <List.Item>
    <List.Icon>
      <RemoveNotificationMutation
        optimisticResponse={{
          removeNotification: notification.id
        }}
      >
        {(removeNotification, { loading }) =>
          loading ? (
            <Icon name="hourglass half" color="grey" title={context.i18n`Removing`} />
          ) : (
            <ClickableIcon
              onClick={() => removeNotification({ variables: { id: notification.id } })}
              name="eye"
              color="grey"
              title={context.i18n`Mark as viewed`}
            />
          )
        }
      </RemoveNotificationMutation>
    </List.Icon>
    <List.Content>
      <div key="1">
        <Time>{context.Ui.relativeDate(notification.createdAt)}</Time> &middot;{' '}
        <Link to={`/process/${context.Ui.urlName(notification.processInstance.process.name)}`}>
          {notification.text}
        </Link>
      </div>
    </List.Content>
  </List.Item>
));
