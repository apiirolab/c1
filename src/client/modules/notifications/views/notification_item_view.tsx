import * as dayjs from 'dayjs';
import * as React from 'react';


// import relativeTime from 'dayjs/plugin/relativeTime'
import { List } from 'semantic-ui-react';
import styled from 'styled-components';

// dayjs.extend(relativeTime);

const Time = styled.span`
  color: #aaa;
`;

type Props = {
  notification: Corpix.Entities.Notification;
};

export const NotificationView: React.SFC<Props> = ({ notification }) => (
  <List.Item>
    <Time>{ dayjs(notification.date).toString() }</Time> &middot; { notification.comment }
  </List.Item>
);
