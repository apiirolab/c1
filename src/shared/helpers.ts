interface Indexable<T> {
  [index: string]: T;
}

interface Group<T> {
  key: string;
  values: T[];
}

export function groupBy<T>(xs: T[], key: string): Indexable<T[]> {
  return xs.reduce(function(previous: any, current: Indexable<any>) {
    // Indexable<Array<Indexable<T>>>
    (previous[current[key]] = previous[current[key]] || []).push(current);
    return previous;
  }, {});
}

export function groupByArray<T>(xs: T[], key: string | Function): Array<Group<T>> {
  return xs.reduce(function(previous, current: Indexable<any>) {
    let v = key instanceof Function ? key(current) : current[key];
    let el = previous.find(r => r && r.key === v);
    if (el) {
      el.values.push(current);
    } else {
      previous.push({
        key: v,
        values: [current]
      });
    }
    return previous;
  }, []);
}
