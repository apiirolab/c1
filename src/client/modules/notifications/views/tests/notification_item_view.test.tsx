import * as React from 'react';

import { create as render } from 'react-test-renderer';

import { create } from 'shared/test_data';
import { NotificationView } from '../notification_item_view';

describe('Notifications', () => {
  storyOf(
    'Item View',
    {
      componentWithData(notification: Corpix.Entities.Notification) {
        return <NotificationView notification={notification} />;
      }
    },
    data => {
      it('renders default view', () => {
        const root = render(
          data.componentWithData(create.notification({ comment: 'This is a comment' }))
        );
        expect(root).toMatchSnapshot();

        const root1 = render(
          data.componentWithData(create.notification({ comment: 'This is another comment' }))
        );
        expect(root1).toMatchSnapshot();
      });
    }
  );
});
