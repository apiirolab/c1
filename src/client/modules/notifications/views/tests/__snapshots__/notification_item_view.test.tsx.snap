// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Item View renders default view 1`] = `
<div>
  <div
    className="item"
    onClick={[Function]}
    role="listitem"
  >
    <span
      className="sc-bdVaJa cWCCxy"
    >
      in 7 months
    </span>
     Â· 
    ProcessStarted
  </div>
</div>
`;
